homeassistant/components/brother/sensor.py

```diff
@@ -0,0 +1,135 @@
+"""Support for the Brother service."""
+import logging
+
+from homeassistant.helpers import device_registry
+from homeassistant.helpers.entity import Entity
+
+from .const import (
+    ATTR_DRUM_COUNTER,
+    ATTR_DRUM_REMAINING_LIFE,
+    ATTR_DRUM_REMAINING_PAGES,
+    ATTR_ICON,
+    ATTR_LABEL,
+    ATTR_MANUFACTURER,
+    ATTR_UNIT,
+    CONF_SENSORS,
+    DOMAIN,
+    SENSOR_TYPES,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Add Brother entities from a config_entry."""
+    brother = hass.data[DOMAIN][config_entry.entry_id]
+
+    sensors = []
+    sensors_list = []
+
+    if brother.available:
+        name = brother.model
+        for sensor in SENSOR_TYPES:
+            if sensor in brother.data:
+                sensors_list.append(sensor)
+        device_info = {
+            "identifiers": {(DOMAIN, brother.serial)},
+            "name": brother.model,
+            "manufacturer": ATTR_MANUFACTURER,
+            "model": brother.model,
+            "sw_version": brother.firmware,
+        }
+    else:
+        _LOGGER.info("Printer is unavailable, reading device info from registry.")
```

Comment:
Interesting approach!  We normally fail setup if the device isn't available at this point in time.  @MartinHjelmare and @balloob what do you think about this?

Reply:
It's weird that we couple the availability status of the device with the source of the device info. The device info is static while the availability status is dynamic. We might very well already have saved the device info before the device went offline.

Maybe we should always fetch the device info from the device registry?

If we don't need any other info from the online device we can continue and add sensors for it, and show them as not available. The alternative is to raise `PlatformNotReady` to try again later.

Reply:
👍 for raising `PlatformNotReady` and letting HA's logic retry.

Reply:
@MartinHjelmare I think raise `PlatformNotReady` in this case isn't good because a lot of people turn off their printer when they don't use it, don't ask me why. We can always read device info from registry... but what with change printer's firmware version? When component should update device info in registry?
Because every model of the printer has another list of sensors component adds to the registry sensor list. I'm not sure that is a good solution. Please look [here](https://github.com/home-assistant/home-assistant/blob/7fe09917e5285637497c05df0b1e1de1449131f7/homeassistant/components/brother/config_flow.py#L66-L68).

Reply:
The device_info is only read when adding the entity. After that it's considered static for the lifetime of the entity. So there's no point in updating it after adding the entity.

Reply:
@MartinHjelmare Yes, I know that. If we always read `device_info` from the registry, the change of the printer firmware version willn't be reflected in the `device_info` in the registry. Only removing and re-adding integration will change the version number.

Reply:
We could wait with forwarding the entry to the platform until we have the first successful device update. Then we know that the device info will have been updated and stored before we set up the platform and add entities.

This will allow a restart to update the device info.

Reply:
@MartinHjelmare @andrewsayre Now integration raises `ConfigEntryNotReady` when printer is offline/unaccessible at HA start.